syntax = "proto3";

import "google/protobuf/duration.proto";
import "scalapb/scalapb.proto";

option (scalapb.options) = {
  flat_package: true
  scope: FILE
  preserve_unknown_fields: false
  package_name: "co.firstorderlabs.common.protos.events"
};

enum DoneReason {
    notDone = 0;
    canceled = 1;
    filled = 2;
}

enum Liquidity {
    global = 0;
    maker = 1;
    taker = 2;
}

enum OrderSide {
    buy = 0;
    sell = 1;
}

enum OrderStatus {
    done = 0;
    open = 1;
    pending = 2;
    received = 3;
    rejected = 4;
}

enum RejectReason {
    notRejected = 0;
    fundsTooLarge = 1;
    fundsTooSmall = 2;
    insufficientFunds = 3;
    postOnly = 4;
    priceTooLarge = 5;
    priceTooSmall = 6;
    sizeTooLarge = 7;
    sizeTooSmall = 8;
}

message BuyLimitOrder {
    option (scalapb.message).extends = "co.firstorderlabs.common.types.Events.BuyOrderEvent";
    option (scalapb.message).extends = "co.firstorderlabs.common.types.Events.LimitOrderEvent";
    option (scalapb.message).extends = "co.firstorderlabs.common.types.Events.SpecifiesSize";

    string orderId = 1 [(scalapb.field).type = "co.firstorderlabs.common.types.Types.OrderId"];
    OrderStatus orderStatus = 2;
    string price = 3 [(scalapb.field).type = "co.firstorderlabs.common.currency.Configs.ProductPrice"];
    string productId = 4 [(scalapb.field).type = "co.firstorderlabs.common.types.Types.ProductId"];
    OrderSide side = 5;
    string size = 6 [(scalapb.field).type = "co.firstorderlabs.common.currency.Configs.ProductPrice.ProductVolume"];
    string time = 7 [(scalapb.field).type = "java.time.Instant"];
    RejectReason rejectReason = 8;
    string requestId = 9 [(scalapb.field).type = "co.firstorderlabs.common.types.Types.OrderRequestId"];
    string doneAt = 10 [(scalapb.field).type = "java.time.Instant"];
    DoneReason doneReason = 11;
    MatchEvents matchEvents = 12;
    google.protobuf.Duration timeToLive = 13 [(scalapb.field).type = "java.time.Duration"];
}

message BuyMarketOrder {
    option (scalapb.message).extends = "co.firstorderlabs.common.types.Events.BuyMarketOrderEvent";
    option (scalapb.message).extends = "co.firstorderlabs.common.types.Events.SpecifiesFunds";

    string funds = 1 [(scalapb.field).type = "co.firstorderlabs.common.currency.Configs.ProductPrice.QuoteVolume"];
    string orderId = 2 [(scalapb.field).type = "co.firstorderlabs.common.types.Types.OrderId"];
    OrderStatus orderStatus = 3;
    string productId = 4 [(scalapb.field).type = "co.firstorderlabs.common.types.Types.ProductId"];
    OrderSide side = 5;
    string time = 6 [(scalapb.field).type = "java.time.Instant"];
    RejectReason rejectReason = 7;
    string requestId = 8 [(scalapb.field).type = "co.firstorderlabs.common.types.Types.OrderRequestId"];
    string doneAt = 9 [(scalapb.field).type = "java.time.Instant"];
    DoneReason doneReason = 10;
    MatchEvents matchEvents = 12;
}

message Cancellation {
    option (scalapb.message).extends = "co.firstorderlabs.common.types.Events.CancellationEvent";

    string orderId = 1 [(scalapb.field).type = "co.firstorderlabs.common.types.Types.OrderId"];
    string price = 2 [(scalapb.field).type = "co.firstorderlabs.common.currency.Configs.ProductPrice"];
    string productId = 3 [(scalapb.field).type = "co.firstorderlabs.common.types.Types.ProductId"];
    OrderSide side = 4;
    string remainingSize = 6 [(scalapb.field).type = "co.firstorderlabs.common.currency.Configs.ProductPrice.ProductVolume"];
    string time = 7 [(scalapb.field).type = "java.time.Instant"];
}

message Order {
    oneof sealed_value {
        BuyLimitOrder buyLimitOrder = 1;
        BuyMarketOrder buyMarketOrder = 2;
        SellLimitOrder sellLimitOrder = 3;
        SellMarketOrder sellMarketOrder = 4;
    }
}

message Orders {
    map<string, Order> orders=1;
}

message Match {
    option (scalapb.message).extends = "co.firstorderlabs.common.types.Events.MatchEvent";

    Liquidity liquidity = 1;
    string makerOrderId = 2 [(scalapb.field).type = "co.firstorderlabs.common.types.Types.OrderId"];
    string price = 3 [(scalapb.field).type = "co.firstorderlabs.common.currency.Configs.ProductPrice"];
    string productId = 4 [(scalapb.field).type = "co.firstorderlabs.common.types.Types.ProductId"];
    OrderSide side = 5;
    string size = 6 [(scalapb.field).type = "co.firstorderlabs.common.currency.Configs.ProductPrice.ProductVolume"];
    string takerOrderId = 7 [(scalapb.field).type = "co.firstorderlabs.common.types.Types.OrderId"];
    string time = 8 [(scalapb.field).type = "java.time.Instant"];
    int64 tradeId = 9 [(scalapb.field).type = "co.firstorderlabs.common.types.Types.TradeId"];
    Order makerOrder = 10;
    Order takerOrder = 11;
}

message MatchEvents {
    repeated Match matchEvents = 1;
}

message SellLimitOrder {
    option (scalapb.message).extends = "co.firstorderlabs.common.types.Events.SellOrderEvent";
    option (scalapb.message).extends = "co.firstorderlabs.common.types.Events.LimitOrderEvent";
    option (scalapb.message).extends = "co.firstorderlabs.common.types.Events.SpecifiesSize";

    string orderId = 1 [(scalapb.field).type = "co.firstorderlabs.common.types.Types.OrderId"];
    OrderStatus orderStatus = 2;
    string price = 3 [(scalapb.field).type = "co.firstorderlabs.common.currency.Configs.ProductPrice"];
    string productId = 4 [(scalapb.field).type = "co.firstorderlabs.common.types.Types.ProductId"];
    OrderSide side = 5;
    string size = 6 [(scalapb.field).type = "co.firstorderlabs.common.currency.Configs.ProductPrice.ProductVolume"];
    string time = 7 [(scalapb.field).type = "java.time.Instant"];
    RejectReason rejectReason = 8;
    string requestId = 9 [(scalapb.field).type = "co.firstorderlabs.common.types.Types.OrderRequestId"];
    string doneAt = 10 [(scalapb.field).type = "java.time.Instant"];
    DoneReason doneReason = 11;
    MatchEvents matchEvents = 12;
    google.protobuf.Duration timeToLive = 13 [(scalapb.field).type = "java.time.Duration"];
}

message SellMarketOrder {
    option (scalapb.message).extends = "co.firstorderlabs.common.types.Events.SellMarketOrderEvent";
    option (scalapb.message).extends = "co.firstorderlabs.common.types.Events.SpecifiesSize";

    string orderId = 1 [(scalapb.field).type = "co.firstorderlabs.common.types.Types.OrderId"];
    OrderStatus orderStatus = 2;
    string productId = 3 [(scalapb.field).type = "co.firstorderlabs.common.types.Types.ProductId"];
    OrderSide side = 4;
    string size = 5 [(scalapb.field).type = "co.firstorderlabs.common.currency.Configs.ProductPrice.ProductVolume"];
    string time = 6 [(scalapb.field).type = "java.time.Instant"];
    RejectReason rejectReason = 7;
    string requestId = 8 [(scalapb.field).type = "co.firstorderlabs.common.types.Types.OrderRequestId"];
    string doneAt = 9 [(scalapb.field).type = "java.time.Instant"];
    DoneReason doneReason = 10;
    MatchEvents matchEvents = 12;
}